cpp/include/cuspatial/detail/utility/linestring.cuh

```diff
@@ -129,5 +136,86 @@ __forceinline__ T __device__ squared_segment_distance(vec_2d<T> const& a,
   return segment_distance_no_intersect_or_colinear(a, b, c, d);
 }
 
+/**
+ * @internal
+ * @brief Given two collinear or parallel segments, return their potential overlapping segment
+ *
+ * @p a, @p b, @p c, @p d refer to end points of segment ab and cd.
+ * @p center is the geometric center of the segments, used to decondition the coordinates.
+ *
+ * @return optional end points of overlapping segment
+ */
+template <typename T>
+__forceinline__ thrust::optional<segment<T>> __device__
+collinear_or_parallel_overlapping_segments(vec_2d<T> const& a,
+                                           vec_2d<T> const& b,
+                                           vec_2d<T> const& c,
+                                           vec_2d<T> const& d,
+                                           vec_2d<T> const& center = vec_2d<T>{})
+{
+  auto ab = b - a;
+  auto ac = c - a;
+
+  // Parallel
+  if (det(ab, ac) != 0) return thrust::nullopt;
+
+  // Must be on the same line, test if intersect
+  if ((a < c && c < b) || (a < d && d < b)) {
+    // Compute smallest interval between the segments
+    auto a_ = a, b_ = b, c_ = c, d_ = d;
```

Comment:
How about just making this function pass `a, b, c, d` by value instead?

Reply:
I think the code is cleaner using pass by value.